src/connection.rs

```diff
@@ -20,59 +20,297 @@
+        connection: &RTCPeerConnection,
+        data_channel_tx: Sender<Arc<DetachedDataChannel>>,
+    ) {
+        connection
+            .on_data_channel(Box::new(move |data_channel: Arc<RTCDataChannel>| {
+                debug!(
+                    "Incoming data channel '{}'-'{}'",
+                    data_channel.label(),
+                    data_channel.id()
+                );
+                let data_channel = data_channel.clone();
+
+                let mut data_channel_tx = data_channel_tx.clone();
+                Box::pin(async move {
+                    data_channel
+                        .on_open({
+                            let data_channel = data_channel.clone();
+                            Box::new(move || {
+                                debug!(
+                                    "Data channel '{}'-'{}' open",
+                                    data_channel.label(),
+                                    data_channel.id()
+                                );
+
+                                Box::pin(async move {
+                                    let data_channel = data_channel.clone();
+                                    match data_channel.detach().await {
+                                        Ok(detached) => {
+                                            if let Err(e) = data_channel_tx.try_send(detached) {
+                                                // This can happen if the client is not reading
+                                                // events (using `poll_event`) fast enough, which
+                                                // generally shouldn't be the case.
+                                                error!("Can't send data channel: {}", e);
+                                            }
+                                        },
+                                        Err(e) => {
+                                            error!("Can't detach data channel: {}", e);
+                                        },
+                                    };
+                                })
+                            })
+                        })
+                        .await;
+                })
+            }))
+            .await;
+    }
 }
 
-impl AsyncRead for Connection<'_> {
-    fn poll_read(
-        mut self: Pin<&mut Self>,
+impl<'a> StreamMuxer for Connection {
+    type Substream = PollDataChannel;
+    type OutboundSubstream = BoxFuture<'static, Result<Arc<DetachedDataChannel>, Self::Error>>;
+    type Error = io::Error;
+
+    fn poll_event(
+        &self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<StreamMuxerEvent<Self::Substream>, Self::Error>> {
+        let mut rx = self.inner.incoming_data_channels_rx.lock().unwrap();
+        match ready!(rx.poll_next(cx)) {
+            Some(detached) => {
+                let ch = PollDataChannel::new(detached);
+
+                let mut channels = self.inner.data_channels.lock().unwrap();
```

Comment:
It feels a bit dangerous to me to lock a mutex here while the `rx` mutex is still locked. Can you not drop the `rx` lock before?